templates/cc/string.go

```diff
@@ -4,14 +4,68 @@ const strTpl = `
 	{{ $f := .Field }}{{ $r := .Rules }}
 	{{ template "const" . }}
 	{{ template "in" . }}
-	{{ if $r.Pattern }}
+	{{ if $r.Len }}
+		if utf8.RuneCountInString({{ accessor . }}) != {{ $r.GetLen }} {
+			return {{ err . "value length must be " $r.GetLen " runes" }}
+		}
+	{{ else if $r.MinLen }}
 		{{ unimplemented }}
-		{{/* TODO(akonradi) implement regular expression constraints.
-		if !{{ lookup $f "Pattern" }}.MatchString({{ accessor . }}) {
-			return {{ err . "value does not match regex pattern " (lit $r.GetPattern) }}
+		{{/* TODO(akonradi) implement UTF-8 length constraints
+		{{ if $r.MaxLen }}
+			{{ if eq $r.GetMinLen $r.GetMaxLen }}
+				if utf8.RuneCountInString({{ accessor . }}) != {{ $r.GetMinLen }} {
+					return {{ err . "value length must be " $r.GetMinLen " runes" }}
+				}
+			{{ else }}
+				if l := utf8.RuneCountInString({{ accessor . }}); l < {{ $r.GetMinLen }} || l > {{ $r.GetMaxLen }} {
+					return {{ err . "value length must be between " $r.GetMinLen " and " $r.GetMaxLen " runes, inclusive" }}
+				}
+			{{ end }}
+		{{ else }}
+			if utf8.RuneCountInString({{ accessor . }}) < {{ $r.GetMinLen }} {
+				return {{ err . "value length must be at least " $r.GetMinLen " runes" }}
+			}
+		{{ end }}
+		*/}}
+	{{ else if $r.MaxLen }}
+		{{ unimplemented }}
+		{{/* TODO(akonradi) implement UTF-8 length constraints
+		if utf8.RuneCountInString({{ accessor . }}) > {{ $r.GetMaxLen }} {
+			return {{ err . "value length must be at most " $r.GetMaxLen " runes" }}
 		}
 		*/}}
 	{{ end }}
+
+	{{ if $r.LenBytes }}
+		const auto length = {{ accessor . }}.size();
```

Comment:
Please enclose this block in braces to limit the scope of `length`. That helps prevent name conflicts.

Reply:
Oh, detail!